Implement support for commit comments

banga opened the pull request:
Since the webhook payload doesn't provide the commit title, which we use
for the push message subject lines, I had to add an API call to fetch
the title.

This _should_ cause gmail to thread commits with comments properly, but
I wasn't able to get sendmail to actually send emails to myself to test
it out.

Still need to split the push email into one email per commit to make
threading perfect.

This is how the e-mail looks in the test harness:
![image](https://cloud.githubusercontent.com/assets/541603/16479319/5e959b82-3e54-11e6-9999-e19b008c113b.png)

Commit:
Implement support for commit comments

Since the webhook payload doesn't provide the commit title, which we use
for the push message subject lines, I had to add an API call to fetch
the title.

This *should* cause gmail to thread commits with comments properly, but
I wasn't able to get sendmail to actually send emails to myself to test
it out.

Still need to split the push email into one email per commit to make
threading perfect.

mihaip commented on app/bettermail.go:
```diff
@@ -127,6 +135,72 @@ func handlePushPayload(payload PushPayload, c appengine.Context) (*mail.Message,
+	commitSHA := *payload.Comment.CommitID
+	url := strings.Replace(*payload.Repo.CommitsURL, "{/sha}", "/" + commitSHA, 1)
+
+	client := urlfetch.Client(c)
```
This won't work for private repos. I would rather not have to make the app more complex by requiring it to store auth tokens that allow read access to repos.

As an alternative design, since you should always get a `push` event before a `commit-comment` event, you could store a simple map of commit ID -> email subject. Then when a comment arrives, you would know what subject to generate. This would also let you avoid splitting multi-commit pushes into separate emails, since you would know what subject to generate for them (generating one email per commit would be annoying for merge pushes that include lots of commits).

mihaip commented on app/bettermail.go:
```diff
@@ -127,6 +135,72 @@ func handlePushPayload(payload PushPayload, c appengine.Context) (*mail.Message,
+		recipient = "mihai@quip.com"
+	}
+
+	message := &mail.Message{
```
If possible, we should start using the `Message-ID` and `In-Reply-To` headers (in addition to subjects) to ensure that we get correct threading in all clients (Gmail threads by subject, but other mail clients may be stricter and use headers only).

banga replied:
Looks like setting `Message-ID` is not allowed: https://cloud.google.com/appengine/docs/go/mail/mail-with-headers-attachments, which means we can't set `In-Reply-To` correctly either.

mihaip replied:
Do you get the message ID after sending? We could persist that along with the subject.

banga replied:
Doesn't look like it: https://github.com/golang/appengine/blob/master/mail/mail.go#L77

mihaip commented on app/bettermail.go:
```diff
@@ -127,6 +135,72 @@ func handlePushPayload(payload PushPayload, c appengine.Context) (*mail.Message,
+		"Sender":                   payload.Sender,
+		"Repo":                     payload.Repo,
+		"ShortSHA":                 commitShortSHA,
+		"Body":                     *payload.Comment.Body,
```
Would be nice to do markdown rendering of the comment body. We do this for commit bodies already, see `newDisplayCommit` for an example.

Commit:
Store e-mail subjects for pushed commits

Had to change some imports because the old style has been deprecated
(https://github.com/golang/appengine#2-update-import-paths)

Commit:
Use markdown rendering on commit comments

mihaip commented on app/bettermail.go:
```diff
@@ -183,24 +193,28 @@ func handleCommitCommentPayload(payload CommitCommentPayload, c appengine.Contex
+	if appengine.IsDevAppServer() {
+		return "mihai@quip.com"
+	}
+	return "shrey@quip.com"
```
Should change this back to `eng+commits@quip.com`

mihaip commented on app/bettermail.go:
```diff
@@ -24,33 +26,62 @@ func init() {
+	    CommitSHA: sha,
+	    Subject: subject,
+	}
+	key := datastore.NewIncompleteKey(c, "EmailThread", nil)
```
You can use `sha` as the key (`datastore.newKey(c, "EmailThread", sha, ...)`, and then you can use `datastore.Get` to more directly look it up.

mihaip replied:
See https://github.com/mihaip/slack-archive/blob/master/app/account.go for an example.

mihaip commented on app/bettermail.go:
```diff
@@ -24,33 +26,62 @@ func init() {
+}
+
+type EmailThread struct {
+	CommitSHA string
```
Should use `datastore:",noindex"` to avoid App Engine indexing these properties (which adds extra overhead).

Commit:
Better datastore usage

Commit:
Fix recipient

mihaip commented:
Thanks! Deployed it.

banga commented:
Woohoo!

mihaip commented:
Looks like commit-level comments (not attached to a specific line) end up with a broken-looking link, do you mind hiding it in that case?

![image](https://cloud.githubusercontent.com/assets/513813/16644756/0e9a8590-43d4-11e6-9234-d7b08df79fcf.png)

banga commented:
Fixed in #2
